Publicize: Allow publishing a post as a Twitter thread.

pento opened the pull request:
# Project Goal

The goal of this project is to create a seamless method for fully featured posts to be transformed automatically into Twitter Threads, fitting into the Twitter format as naturally as possible. Along the way, it should provide useful feedback to the author, indicating how their post will be transformed, and giving them the opportunity to tweak their post content based on how they would like the thread to appear.

# Overviews

For an overview of the logic behind mapping post content to tweets, I recommend reading the [Content Overview](https://github.com/Automattic/jetpack/pull/16699#issuecomment-690009821).

If you're after technical information about how all of the parts work together, then you'll want to head over to the [Technical Overview](https://github.com/Automattic/jetpack/pull/16699#issuecomment-690010800).

If you want to learn how you can most effectively help with testing, and what you can look for to indicate incorrect behaviour that should be reported, check out the [Product Testing Overview](https://github.com/Automattic/jetpack/pull/16699#issuecomment-693176464).

Finally, if there's a feature missing that you feel should be included, please check the [Future Work Overview](https://github.com/Automattic/jetpack/issues/17149) first, as there are [alot](//core.trac.wordpress.org/ticket/50000) of things I'd love to include, but not everything will make it into the first release.

# Screenshots

<img width="400" alt="" src="https://user-images.githubusercontent.com/352291/90095597-7a898880-dd74-11ea-952b-e090d0bc8ed5.png"> <img width="400" alt="" src="https://user-images.githubusercontent.com/352291/90365128-bab87600-e0a8-11ea-8a46-8a404ec763ca.png">
 <img width="400" alt="" src="https://user-images.githubusercontent.com/352291/90364190-5b0d9b00-e0a7-11ea-803f-a30a904d765c.png"> <img width="400" alt="" src="https://user-images.githubusercontent.com/352291/91515253-1e099a00-e92c-11ea-91a5-89fa8bc45019.png"> <img width="400" alt="" src="https://user-images.githubusercontent.com/352291/91515250-1c3fd680-e92c-11ea-959a-3b3215d9ed3c.png"> <img width="400" alt="" src="https://user-images.githubusercontent.com/352291/91515244-177b2280-e92c-11ea-9210-0927832a8ee2.png">

# Housekeeping

#### Changes proposed in this Pull Request:

* Adds a new feature: Publishing Twitter threads from within the block editor, via Publicize.

#### Jetpack product discussion

p1HpG7-9US-p2

#### Does this pull request change what data or activity we track or use?

No change, it only uses existing data (eg, Keyring for Twitter connection, post content for generating thread)

#### Testing instructions:

##### Setup

If you need to test parser changes on your sandbox:

- Apply the relevant diff to your sandbox.
- Ensure you've added `add_filter( 'publicize_force_synchronous', '__return_true' );` on your sandbox.
- Define `JETPACK__SANDBOX_DOMAIN` to your sandbox domain in the `wp-config.php` file on your Jetpack site.

##### Testing

Please see the [Product Testing Overview](https://github.com/Automattic/jetpack/pull/16699#issuecomment-693176464) details on testing.

#### Proposed changelog entry for your changes:

* Publicize: Add the ability to publish the entire content of posts to Twitter as threads.

pento commented:
@Automattic/jetpack-crew: Could I get your thoughts on the hackery I've had to do `packages/autoloader/src/AutoloadGenerator.php` in order to be able to load the `nojimage/twitter-text-php` Composer package? The fundamental problem appears to be that `nojimage/twitter-text-php` specifies a `PSR-0` autoloader, but not a `PSR-4` autoloader: it needs someone more familiar with these PSRs to determine if what I've done is correct. I'm happy to delete all the code if there's a better way. 🙂

Also, `nojimage/twitter-text-php` requires the PHP `intl` extension to be installed, I've added that to the Dockerfile, you might need to do a `yarn docker:build-image` if you're testing there. They suggested using `symfony/polyfill` to provide support for PHP installs that don't have `intl` installed, but I wasn't able to get that working.

jeherve commented:
> The fundamental problem appears to be that nojimage/twitter-text-php specifies a PSR-0 autoloader, but not a PSR-4 autoloader: it needs someone more familiar with these PSRs to determine if what I've done is correct.

cc @kbrown9 

> Also, nojimage/twitter-text-php requires the PHP intl extension to be installed

Do you foresee that being a problem once we release the block to the public? Is the number of configs not including `intl` out of the box high enough that we need to ensure a graceful fallback?

I'm asking because although it is a pretty standard package, after a quick look around, it would seem that some popular hosts may not bundle it by default with their old PHP versions ([example](https://help.dreamhost.com/hc/en-us/articles/360003820252-How-to-enable-the-intl-extension)).

pento commented:
> Do you foresee that being a problem once we release the block to the public? Is the number of configs not including `intl` out of the box high enough that we need to ensure a graceful fallback?
> 
> I'm asking because although it is a pretty standard package, after a quick look around, it would seem that some popular hosts may not bundle it by default with their old PHP versions ([example](https://help.dreamhost.com/hc/en-us/articles/360003820252-How-to-enable-the-intl-extension)).

That's something I'm not sure about. 🙂 I didn't know if we had any data on what modules folks have enabled when they're running Jetpack.

There are a few alternatives to look into:
- Figure out how to use the `symfony/polyfill` package, but I didn't want to get into that until I'd confirmed the right way to include packages.
- Contribute a patch to `nojimage/twitter-text-php` that allows some sort of fallback when `intl` isn't available.
- Do the fallback in Jetpack: worst case is that the tweet parsing is done incorrectly on the Jetpack site, but the tweets which are ultimately published to Twitter will be done on WPCOM.

pento commented:
@kbrown9: Hey, could I get your thoughts on the autoloader problem mentioned earlier?

pento commented:
@Automattic/jetpack-crew: FYI, I've removed the `twitter-text-php` dependency, so the Autoloader changes (and `php-intl` dependency) are no longer required. 🙂

kbrown9 commented:
> @kbrown9: Hey, could I get your thoughts on the autoloader problem mentioned earlier?

Sorry for the delay - I missed the earlier message about the PSR-0 autoloader! I see that the issue is resolved now :smile:

jeherve reviewed (commented):
This should be super nice, and it works nicely already! 🤩 

In terms of user experience, so far my biggest comment would be about the post / thread title:
- I would have expected the post title and / or the Publicize custom message to be the first messages in the Tweetstorm, as a bit of an introduction to the thread.
- I wonder if the first tweet of the thread should somehow indicate that this is the thread: on Twitter some folks use "1/x", others use "⬇". Should we add something similar to the first tweet?

pento commented on extensions/blocks/publicize/twitter/editor.scss:
```diff
@@ -0,0 +1,121 @@
+	.jetpack-publicize-twitter__tweet-divider-icon {
+		margin: 0 auto;
+		display: block;
+		background: #0009;
```
Nope, the use of black-with-transparency (and white-with-transparency for dark themes) is intentional: the UI needs to be visible against any background colour the theme might set in the editor. Gutenberg itself handles this by providing starkly contrasting UI: it's all black and white, with appropriate borders for each. The Twitter UI is intended to be more subtle, so that it can remain visible, without distracting from the content itself.

For interest, I wrote up an experiment to evaluate the accessibility and visual appearance of different backgrounds against a large range of background colours.

https://jsfiddle.net/pento/83kd0na5/2/

pento commented on _inc/lib/class-jetpack-tweetstorm-helper.php:
```diff
@@ -8,12 +8,135 @@
+	 *
+	 * @var string
+	 */
+	private static $line_seperator = "\xE2\x80\xA8";
```
Well spotted, thanks! Fixed in 792cb5c97.

pento commented on _inc/lib/class-jetpack-tweetstorm-helper.php:
```diff
@@ -8,12 +8,135 @@
+	 * @var array
+	 */
+	private static $supported_blocks = array(
+		'core/gallery'   => array(
```
Yep, I still need to audit all of the Jetpack blocks to ensure everything that can be supported, is.

pento commented:
Thanks for the review, @jeherve! I appreciate you volunteering as tribute. 😉

> This should be super nice, and it works nicely already! 🤩

Great to hear it worked well for you! It's been shaping up pretty well, I'm happy with how it's going. 🙂

> I would have expected the post title and / or the Publicize custom message to be the first messages in the Tweetstorm, as a bit of an introduction to the thread.

Absolutely, I'm inclined to think the first message would use the Publicize custom message (which defaults to the post title), perhaps with a "this is a thread" indicator, as you suggested. Additionally, the last message could be a "this thread was also published on my site" thing, with a link to the published post. 

> I wonder if the first tweet of the thread should somehow indicate that this is the thread: on Twitter some folks use "1/x", others use "⬇". Should we add something similar to the first tweet?

There's quite a variety of ways that people indicate threads: both the notation at the beginning of the thread, as well as how tweets within the thread are marked. I haven't explored the best way to provide this functionality yet, though I think the MVP for it would be relatively straightforward.

marekhrabe commented:
I was testing just the preview part of this and I have noticed there is a big difference of the empty state (no title and no text body) between the "single link" and "thread" modes:

| Single Tweet | Thread |
| - | -
| <img width="630" alt="Screenshot 2020-09-08 at 14 48 59" src="https://user-images.githubusercontent.com/156676/92480695-8f2b3600-f1e5-11ea-925a-9f0a7a10f8eb.png"> | <img width="809" alt="Screenshot 2020-09-08 at 14 49 22" src="https://user-images.githubusercontent.com/156676/92480709-93575380-f1e5-11ea-8764-da8c653cdd13.png">

Can we do something nicer for the thread one? It looks really good when filled with data, just the empty one is strangely blank

pento commented:
Good catch, thanks @marekhrabe!

I'm planning on adding an intro tweet to the start of the thread (probably based on the post title/custom Publicize message), and a closing tweet at the end of the thread, linking to the post: these will hopefully fill in the blank space, but I'll make sure to keep it in mind as I'm working on it. 🙂

pento commented:
Hey, this PR is finally in a state that's worth reviewing! I'm currently working on a comprehensive overview for how it all works, I expect to have it finished in the next day or two.

Please don't feel alarmed at the size of this PR: I know there's a lot of code, and it can be hard to figure out where to start. I'm intending to cover it in the overview, which will make it easier to see where the boundaries of each section of the code is, as well as how (and why) things function within each section.

In the mean time, please enjoy this GIF as a not-so-subtle metaphor for what this project has turned into.

![](https://thumbs.gfycat.com/BowedCourageousCommabutterfly-size_restricted.gif)

pento commented:
# Content Overview

Broadly speaking, Twitter supports three types of content: text, media, and embeds. There are some subdivisions and interdependencies between these types, but it's useful to keep these broad categories in mind. When looking to transform a block into a tweet, the content of that block needs to fit into at least one of these categories: if it doesn't, it's not a block that we can support.

## Text Content

Twitter only allows plain text content, but there are some conventions which allows for limited formatting to apply to certain kinds of text. For example, it's common to see each item in a list having a hyphen for a prefix.

Twitter's 280 character limit makes it challenging to directly map blocks to tweets, since a paragraph can be a tiny fraction of that, or can be several times that size. While a human author can edit their content to fit those limits in a natural way, it's challenging to automate. The most complex parts of this PR exist to address these challenges.

## Media Content

A tweet can have one GIF, one video, or four still images attached to it. While gallery blocks are significantly more flexible than that, it's not particularly common to see GIFs or videos used in them. Galleries are most often still images, while GIFs will most often be inserted as single image blocks, and videos will be inserted as video blocks. This does allow us to simplify the mapping a lot: single image and video blocks map directly to a tweet, while gallery blocks can be filtered down to the first four still images.

## Embed Content

When a URL is placed in the text of the tweet, Twitter will attempt to create a "card", a visual representation of the URL, based on the OpenGraph data that can be extracted from that URL, much like how WordPress uses oEmbed. While a single tweet can only ever have one card, it does mean that mapping embed blocks to a tweet is relatively straightforward.

pento commented:
# Technical Overview

This project can be broadly split into four major parts: the block editor UI, the tweet parser, the thread preview UI, and the WPCOM Publicize support for threads. Only the first two are in this PR, so I'll go into the most detail with them here.

I would suggest having the code open in your editor of choice while you read this, so you can refer back and forth more easily.

## Publicize

This can be seen in D48484-code, and is primarily glue code, taking the output of the tweet parser, and feeding it into the existing tweet publishing code. There are more details on the diff of the WPCOM-specific changes.

## Thread Preview UI

This can be seen in Automattic/wp-calypso#44956, and largely consists of expanding the existing social previews feature to display multiple tweets at a time.

## Block Editor UI

The block editor UI is driven entirely by the tweet parser. When relevant content changes, it will send a parse request to the server, which will return an array of tweets, along with information about which blocks map to which tweets, and if blocks needed to be split in some way.

First of all, here's a list of some relevant terms and concepts:
- Block: It's a standard WordPress block that we know and love. A single block can map to zero or more tweets.
- Tweet: It's 280 characters of your best hot takes. A single tweet can map to one or more blocks.
- Content attribute: Blocks have attributes, but not all of them will contribute to how the tweet is constructed. We really only care about regenerating the list of tweets when a relevant "content" attribute changes.
- Boundaries: When a text block is split up into multiple tweets, we need to indicate where the split happens, that location is a boundary. There are two types of boundaries: `normal`, and `end-of-line`. The former is handled using the block editor's [Annotations API](https://developer.wordpress.org/block-editor/developers/block-api/block-annotations/), adding a visual highlight in text, showing where it will be split. The latter is specific to `core/list` blocks, and indicates a break happening at the end of a list item: the Annotations API is unable to handle showing a highlight here, so we have a special case for it. 

With those concepts in mind, let's walk through the lifecycle of how the UI is applied to the block. Starting in `publicize/twitter/index.js`, the `addTweetDivider()` function is run on every block type when it's registered, wrapping its `edit` rendered in the `TweetDivider` component. While this is primarily used for adding data to block types that we do support, it's also used to add a "not supported" message to block types that aren't supported.

`TweetDivider` is defined in `publicize/twitter/tweet-divider.js`, and it's where the bulk of the UI logic lives:

The `render()` method is fairly straightforward, adding the divider UI, any popover warnings we want to show, and the `blockStyles` section is to handle the `end-of-line` type of boundaries.

For updating the UI, the `componentDidUpdate()` method checks to see whether the content attributes of the block have updated, and dispatches the `refreshTweets()` action, (defined in the `withDispatch()` section), which ultimately calls the `__refreshTweets()` function in `publicize/store/effects.js`.

`__refreshTweets()` grabs a copy of all of the blocks that we support from the editor, and sends them off for parsing. The results of that are added to the store, which takes us back to the `withSelect()` section of `TweetDivider`. This code handles checking whether the block is assigned to any tweets, and if so, are there any boundaries that we need to turn into annotations, or popover warnings to show.

And that covers the major steps in the lifecycle of mapping tweets to blocks. There's one notable side-path, however: generating Twitter cards. Twitter cards are used in the social preview UI, when a tweet contains a URL, Twitter will try to generate a card based on the OpenGraph data from that URL, so we do something similar, to provide a more accurate preview UI.

If we go back to `__refreshTweets()`, we can see that it also dispatches the `getTwitterCards()` action, which calls back to the server. The server will attempt to generate the card data for each of the URLs passed, and returns it. While this functionality is editor-specific, we do still have to do the work on the server, due to cross-site headers preventing us from grabbing a copy of the URL from within the browser.

## The Tweet Parser

This is the area of code with the most complex workings: I've cleaned it up a lot, but there are still some bits which can be tricky to follow. The complexity largely comes from two major sources:

- In order to keep things consistent, we need to use the same code to generate the thread in the editor, as we do to generate the thread on WPCOM. This means we're restricted to extracting block data through `parse_blocks()`, which doesn't automatically parse attributes out of the block HTML.
- The biggest restriction when mapping freeform post content to tweets is the character limit. In order to map it in a more natural manner, the parser tries several different methods for splitting long text-blocks into tweets: for example, it tries to split list blocks by list item, then it tries to split by sentence, then it falls back to splitting mid-sentence if there's no alternative.

The tweet parses takes a two phased approach: it first extracts content from all of the blocks it was passed, filtering and formatting it, but not considering whether it fits inside the limitations of a single tweet. The second phase takes that content, and determines how it can be mapped to tweets.

So, lets have a walk through `Jetpack_Tweetstorm_Helper`.

Starting at the top, the `::$supported_blocks` member is a list of all of the blocks we support. The docblock explains what each of the settings is for, but I think they make more sense given the context of where they're used.

It's also worth noting the `::$tweets` member, which is where tweets are stored, as they're being generated. Methods that create, retrieve and update tweets are all doing it through this variable.

Going down the `::parse()` method, this is where you can see the split between the two phases: the first phase happens in the `::extract_blocks()` call, and the second phase happens in the loop that takes up the rest of the `::parse()` method.

### Phase 1: Extraction

Jumping down to `::extract_blocks()`, this is where it loops over the blocks, and attempts to extract relevant data.

Looking at `::extract_text_from_block()` first, it's worth noting that this code currently only supports extracting text content from the block HTML: there isn't a path of extracting text content from serialised block attributes, though adding it would be fairly straightforward.

To give some context for how text is extracted, lets compare two blocks in `::$supported_blocks`: `core/paragraph`, and `core/quote`. Both have their `content_location` set to `html`, but paragraph has its `content` set to an empty array (which is shorthand for "grab all of the text content in the block HTML), whilst quote defines particular tags where the content can be found. The difference here is due to how the quote block is implemented in the editor: it uses two `RichText` components for editing: one for the actual quote, and one for the citation. So, by defining which HTML tags correspond to which chunk of the block, we gain the ability to reverse engineer how the content is represented in the editor itself, as well as giving us more control over how complex blocks are formatted into Twitter-friendly text.

Returning to `::extract_text_from_block()`, we can see that it's using those tags to call `::extract_tag_content_from_html()`, a helper method for extracting the text content of arbitrary tags from an arbitrary blob of HTML. Thankfully, this is much more straightforward than it might sound: WordPress already has a substantial library of battle-tested HTML parsing functions which do all the heavy lifting for us: we just have to decide which parts of the parsed HTML we want to keep.

One thing it does do to help for later is to extract any URLs from the text, and replace them with a placeholder value. URLs take up 24 characters, so by replacing arbitrary length URLs with a fixed length placeholder, we save ourselves a lot of complex re-calculating of tweet length in the mapping phase.

Having extracted the tag content, we come across the first instance of a repeating theme in the parser: treating single line blocks as "multiline" with just one line. The list block is a multiline block: each list item is a separate line, and we apply the block `template` to each line. The paragraph block is a single line block, but if we think of it in terms of being a "multiline" block with just one line, then the same logic applies: we apply the block `template` to each line, it's just that there's only one line.

So, we loop over the tags and lines, apply them to the `template`, do some cleanup, then we return the text in the form we want to use.

`::extract_blocks()` then calls `::extract_media_from_block()`. This is a much simpler extraction, using the `::extract_attr_content_from_html()` method to extract the value of particular attributes on particular tags. Looking at the definition of the `core/gallery` and `core/image` blocks in `::$supported_blocks`, we can see they both define their `content` settings in the same way: the image URL comes from the `src` attribute on the `<img>` tag, and the alt text comes from the `alt` attribute of the `<img>` tag.

Since there's no `template` to apply to these values, it's simply a matter of extracting and returning them.

One thing to note is the special handling for VideoPress in this method: this is a good example of where I'd ultimately like to see custom behaviour handled through a standard WordPress plugin interface: providing appropriate actions and filters for plugin to be able to add their own blocks, or modify how core blocks are treated.

`::extract_tweet_from_block()` gets its own special method, because Twitter allows us to define a "quote" tweet, which is handle differently to how other URLs are turned into cards. So, we extract it separately.

Finally, we look at all other embed-type blocks in `::extract_embed_from_block()`. This method just needs to extract a single URL (which is always stored in a serialised attribute on the block, making it easy to access). Similar to VideoPress in the media extraction, there's some special handling of Jetpack GIF blocks, since the URL needs to be changed for it to work well with Twitter. This is another example of functionality which should be handled through a plugin interface.

### Phase 2: Mapping

There are two factors which drive decisions in the mapping phase: how content fits within the limitations of single tweet, and how it it fits within the context of a Twitter thread.

Starting from the loop in the `::parse()` method, we can see that come into play immediately. Blocks with the "break" type (such as spacer blocks, and divider blocks) force a new tweet to be started, which doesn't make sense in the context of a single tweet, but within a Twitter thread, it fits much more naturally: a divider in the post content is a place for the reader to pause, when reading that same content in a Twitter thread, the break between tweets can offer that same moment of pause.

Looking at `::$supported_blocks` again, we can see some blocks (such as `core/heading`, which have the `force_new` flag set. This flag is for blocks that are best placed at the start of a tweet, headings are good example of that. Conversely, `core/image` has the `force_finished` flag set, which forces the next block to be at the start of a new tweet. It would be confusing if you had a short paragraph, then an image, then another short paragraph in your post, and we squashed the two paragraphs into a one tweet: the reader would see both the paragraphs before the image, which isn't the order that the author intended.

Now, we move on to `::add_text_to_tweets()`. At first glance, this looks like one of the more complex parts of the code, but it largely consists of nested patterns. Much like we did earlier, it starts by treating single line blocks as being multiline, with one line. This allows us to use the same code paths for prioritising how the text is split (if it needs to be).

Next, it loops over each line, and tests to see if that line could be appended to the current tweet. If it can, great! Append, and move to the next line.

If it can't be appended, test to see if that line could fit in a tweet on its own. If it can, great! Start a new tweet with that line, and move on to the next line.

If it can't fit in its own tweet, split the line up into sentences, and loop over each sentence, testing if the sentence can be appended to the current tweet. If it can, great! Append it, and move on to the next sentence.

If it can't be appended, test to see if that sentence could fit in a tweet on its own. If it can, great! Start a new tweet with that sentence, and move on to the next sentence.

If it can't fit in its own tweet, split the sentence up into words, and loop over each word, testing if the word can be appended to the current tweet. If it can, great! Append it, and move on to the next word.

If it can't be appended, start a new tweet with that word, and then move on to the next word.

The devil is in the details, however. 🙂

`::start_new_tweet()`, `::save_current_tweet()`, and `::get_current_tweet()` are all fairly straightforward: they're just helper methods that interact with `::$tweets`.

`::is_valid_tweet()` is a method for testing whether the string that's passed to it fits into Twitter's tweet length limitations. When we need to split a block across multiple tweets, `::get_boundary()` determines the position of the boundary that we need to send back to the editor, for it to display. These two methods have some complexities for their own reasons.

Up until now, we've been keeping track of string lengths using `strlen()`. Rather than saying we're counting characters here, it's a little more accurate to say that we're counting bytes. The primary purpose of counting in this way is for performance, the secondary purpose is consistency: it's easier to count bytes than it is to count multibyte characters.

In order to segment off these sections that work with multibyte characters, any counts that they produce are only used within their sections, even the data that's returned by `::get_boundary()` is only done so for the purpose of storing it, so it can be returned to the editor at the end.

However we do need to consider multibyte characters at some points. `::is_valid_tweet()` calls `::is_within_twitter_length()`, which tries to apply Twitter's character counting rules in a moderately optimised fashion. Generally speaking, these are Twitter character counting rules:

- ASCII characters, as well as characters from certain defined Unicode ranges, count for 1 character.
- Emoji, regardless of whether they consist of a single Unicode character (eg, 🙂), or are constructed as a ZWJ sequence from a bunch of component emoji characters (eg, 👩‍👩‍👧‍👧), count for 2 characters.
- All other characters count for 2 characters.

To optimise this procedure, `::is_within_twitter_length()` tries to deal with the fastest cases first, falling back to the slower cases only when it has to.

`::get_boundary()` has two things it has to do. First of all, we're now in the mapping phase, so all of the content has been formatted for Twitter: where we're splitting it doesn't map directly back to how the content appears in the editor. Items from the list block have had `- ` prepended to them, quote blocks have had curly quotes (“”) added, etc. In order to reverse that, it has to break up the `template` for the block, and determine which parts of it are formatting, and which parts of it are content. Additionally, for blocks with multiple `RichText` areas (ie, the quote block), it needs to determine which one the boundary will occur in.

Once `::get_boundary()` has determined where (in bytes) the boundary occurs, it needs to convert that to a format that the block editor uses. The block editor uses standard JS functions for counting string lengths, which count in UTF-16 code units, so we convert it before returning.

Now that we're done adding the text to the tweet, we can move on to media, in `::add_media_to_tweets()`. This just applies Twitter's rules before attaching the media to a particular tweet:

- Up to four still images per tweet; or
- One GIF per tweet; or
- One video per tweet.

Each media element attached does count for 24 characters, so we also need to check that adding media won't take up too much space.

And then, `add_tweet_to_tweets()` and `add_embed_to_tweets()` are a breath of fresh air in their relative simplicity. 🙂

Having added all the blocks to the tweet, `::clean_return_tweets()` is called, which replaces the URL placeholders with the actual URL they represent, and removes a bunch of our internal flags that the editor doesn't need to know about (substantially shrinking the size of the data blob being returned).

And that's how tweets are parsed from blocks! Both the editor and Publicize follow the same process (with the exception of calls from Publicize skipping `::get_boundary()`, since that's not something it needs to use), which ensures the thread will be posted to Twitter in the same form it appears in the editor.

jeherve commented on extensions/blocks/publicize/twitter/tweet-divider.js:
```diff
@@ -0,0 +1,324 @@
+					return;
+				}
+
+				// Remove any existing annotations in this block.
+				const annotations = select( 'core/annotations' ).__experimentalGetAllAnnotationsForBlock(
+					childProps.clientId
+				);
+				annotations.forEach( annotation => {
+					if ( annotation.source === 'jetpack-tweetstorm' ) {
+						dispatch( 'core/annotations' ).__experimentalRemoveAnnotation( annotation.id );
+					}
+				} );
+
+				const boundaries = tweets.filter( tweet => tweet.boundary ).map( tweet => tweet.boundary );
+
+				// Add new annotations in the appropriate location.
+				boundaries.forEach( boundary => {
+					const { container, type, start, end } = boundary;
+					if ( 'normal' === type ) {
+						dispatch( 'core/annotations' ).__experimentalAddAnnotation( {
```
For my own reference and for anyone who may be asking themselves the same thing; although experimental, those were already available in WordPress 5.4.

jeherve reviewed (commented):
This is really nice, and it works well!

*******

Should we grey those out once a post has been published?

![image](https://user-images.githubusercontent.com/426388/92702602-b141c700-f351-11ea-8430-d76f0122ea5b.png)

*******

If you add a link with the link itself as anchor text, e.g.

```
<p><a href="https://www.instagram.com/p/CCV_I0Rjm6j/">https://www.instagram.com/p/CCV_I0Rjm6j/</a></p>
```

You get something like this on Twitter:
![image](https://user-images.githubusercontent.com/426388/92703679-b18e9200-f352-11ea-8ce9-6dc5449fcacc.png)

Should we try to avoid that?

*******

I had a video block with a VideoPress video in one of my test posts, and it didn't get pulled in the thread. Here is the block's content:

```
<figure class="wp-block-video alignwide wp-block-embed is-type-video is-provider-videopress"><div class="wp-block-embed__wrapper">
https://videopress.com/v/Uvh9ON4T?preloadContent=metadata
</div></figure>
```

pento commented on _inc/lib/core-api/wpcom-endpoints/class-wpcom-rest-api-v2-endpoint-tweetstorm-parse.php:
```diff
@@ -0,0 +1,110 @@
+ * REST API endpoint for parsing Tweetstorms out of block content.
+ *
+ * @package Jetpack
+ * @since 8.7.0
```
👍🏻 8db4fc615

pento commented on extensions/blocks/publicize/store/selectors.js:
```diff
@@ -1,24 +1,257 @@
+
+	return {
+		...getFirstTweet( state ),
+		text: __( "I've also published this thread on my site:", 'jetpack' ) + ` ${ url }`,
```
Ah, I should've documented that. 🙂 The URL _must_ be the last thing in the tweet, so that Twitter will trim it from the tweet text when it displays that tweet.

I've added a comment in b001bf38d to note that.

pento commented:
@jeherve: While looking at the repeated URL problem, I ran into a related problem, which I'd like your thoughts on before moving forward.

The problem is if a URL is typed into the text (not formatted as a link) in a way that Twitter recognises as a URL. For example, if you type "jetpack.com" into the Tweet box on twitter.com, you can see that it automatically turns into a link. All URLs count for 24 characters, which is a problem when URLs shorter than 24 characters are entered in the block editor: they will count for fewer characters than they should, which can cause tweets to be longer than they should be.

The good news is, this can be solved with an existing library: [twitter-text-php](https://github.com/nojimage/twitter-text-php) is up-to-date with Twitter's official [twitter-text](https://github.com/twitter/twitter-text/) library, and is able to detect these URLs. ff990a5 is a hacky test implementation of it, but it works well, I can easily clean it up if we move forward with it.

The bad news is, this was the library that gave me trouble earlier, trying to include as a composer dependency. It really needs to be kept updated through a package manager, since we want to include updates when they're released, but it's probably going to need some sort of changes to the autoloading stuff. Since we don't use any of the parts of the library that require `intl`, we at least get to skip that dependency. 🙂

What are your thoughts on the best way to handle this?

jeherve commented:
> it's probably going to need some sort of changes to the autoloading stuff.

I'll let @kbrown9 weigh in on this; she'll know best. 👍

kbrown9 commented:
> @jeherve: While looking at the repeated URL problem, I ran into a related problem, which I'd like your thoughts on before moving forward.
> 
> The problem is if a URL is typed into the text (not formatted as a link) in a way that Twitter recognises as a URL. For example, if you type "jetpack.com" into the Tweet box on twitter.com, you can see that it automatically turns into a link. All URLs count for 24 characters, which is a problem when URLs shorter than 24 characters are entered in the block editor: they will count for fewer characters than they should, which can cause tweets to be longer than they should be.
> 
> The good news is, this can be solved with an existing library: [twitter-text-php](https://github.com/nojimage/twitter-text-php) is up-to-date with Twitter's official [twitter-text](https://github.com/twitter/twitter-text/) library, and is able to detect these URLs. [ff990a5](https://github.com/Automattic/jetpack/commit/ff990a5f001f600ee1fbf5efecadb281c9bd94fc) is a hacky test implementation of it, but it works well, I can easily clean it up if we move forward with it.
> 
> The bad news is, this was the library that gave me trouble earlier, trying to include as a composer dependency. It really needs to be kept updated through a package manager, since we want to include updates when they're released, but it's probably going to need some sort of changes to the autoloading stuff. Since we don't use any of the parts of the library that require `intl`, we at least get to skip that dependency. 🙂
> 
> What are your thoughts on the best way to handle this?

I can open a PR that adds PSR-0 support to the Jetpack autoloader. I didn't see any problems with how you added PSR-0 support in a previous commit, but it would be nice to have a couple of people take a took at the change. I'll try to get the PR submitted today, and hopefully it can be reviewed early next week. Will that work for you?

Also, I want to make sure that we all understand how the package will be loaded by the Jetpack autoloader. The Jetpack autoloader selects the latest version of a package from the packages that are installed by plugins that use the Jetpack autoloader. If another plugin on a site uses the `nojimage/twitter-text-php` package, and that plugin does not use the Jetpack autoloader (for example, if it uses Composer's autoloader), we can't control which version of the package is loaded on that site.

I don't think this will be a problem with the `nojimage/twitter-text-php` package. It seems unlikely that many plugins are using this package, so version conflicts between plugins that use this package should be rare. I just wanted to make sure we're all aware of this.

pento commented:
Thanks for that, @kbrown9! I think that's a reasonable tradeoff for now, I had [a brief search](https://www.google.com/search?q=extractor%20twitter%20site%3Agithub.com%20inurl%3Awp-plugins) through the plugin directory, and couldn't find anything that's likely to clash.

I'd very much like for this feature to land in Jetpack 9.0, and given the size, it would need to be merged well in advance of code freeze. I'd greatly appreciate the autoloader stuff happening as quickly as possible, so I can get that bit of this PR ticked off. 🙂

pento commented:
> Should we grey those out once a post has been published?
> 
> ![image](https://user-images.githubusercontent.com/426388/92702602-b141c700-f351-11ea-8430-d76f0122ea5b.png)

I'm inclined to allow changing it, since we allow re-publicizing of posts, the author may wish to change how it's published to Twitter.


> If you add a link with the link itself as anchor text, e.g.
> 
> ```
> <p><a href="https://www.instagram.com/p/CCV_I0Rjm6j/">https://www.instagram.com/p/CCV_I0Rjm6j/</a></p>
> ```
> 
> You get something like this on Twitter:
> ![image](https://user-images.githubusercontent.com/426388/92703679-b18e9200-f352-11ea-8ce9-6dc5449fcacc.png)

Fixed in 0df6f95a6.

> I had a video block with a VideoPress video in one of my test posts, and it didn't get pulled in the thread. Here is the block's content:
> 
> ```
> <figure class="wp-block-video alignwide wp-block-embed is-type-video is-provider-videopress"><div class="wp-block-embed__wrapper">
> https://videopress.com/v/Uvh9ON4T?preloadContent=metadata
> </div></figure>
> ```

Aye, I haven't added support for VideoPress embed blocks (ie, VideoPress URL pasted into the editor), only VideoPress video blocks (ie, VideoPress video selected from the media library). I've added it to the list in #17149.

jeherve commented:
> I'm inclined to allow changing it, since we allow re-publicizing of posts, the author may wish to change how it's published to Twitter.

That option is only available to folks with a Jetpack Premium plan though, and via the WordPress.com interface. I'm not sure we can add such an option to wp-admin just yet.

pento commented:
> That option is only available to folks with a Jetpack Premium plan though, and via the WordPress.com interface. I'm not sure we can add such an option to wp-admin just yet.

Aye, that's a fair point. It looks like `RadioControl` doesn't have a `disabled` property, so I'm not sure if there's a nice way of doing this, but I'll see what I can come up with.

pento commented:
# Product Testing Overview

It's worth restating the project goal, since this provides the lens for viewing the product, and seeing how it should function.

> The goal of this project is to create a seamless method for fully featured posts to be transformed automatically into Twitter Threads, fitting into the Twitter format as naturally as possible. Along the way, it should provide useful feedback to the author, indicating how their post will be transformed, and giving them the opportunity to tweak their post content based on how they would like the thread to appear.

Let's break that down, talk about the ideal behaviour of each point, and where it might break.

## "Seamless Method"

The overall flow of the feature should fit naturally into the existing experience of writing a post in the block editor.

### What would a bug look like?

- If the UI gets in the way while you're trying to write the post.
- If the post content is changed in any way.
- If the thread published to Twitter doesn't match expectations.

### How to break it

- Try doing unusual thing with the UI. Interact with it in different ways.

## "Fully Featured Posts"

Within [the limitations of what kinds of content can be posted to Twitter](https://github.com/Automattic/jetpack/pull/16699#issuecomment-690009821), we should support transforming as much  post content as possible into the thread.

### What would a bug look like?

- Content from supported blocks being ignored.
- Unsupported blocks that should be supported.

### How to break it

- Check [the list of blocks that are supported](https://github.com/Automattic/jetpack/pull/16699/files#diff-f5d995c09945daf88f71798f95f5be06R57-R159), and [those which we need to add support for](https://github.com/Automattic/jetpack/issues/17149). Are there any missing? 
- Can you add supported blocks to the post in ways that cause them to be missed?

## "Transformed Automatically"

The transformation from post to thread is entirely automatic, there's no option for user intervention. This means the transformation needs to be reliable, and predictable.

### What would a bug look like?

- If the UI within the block editor doesn't match the social preview, or either of them don't match the tweeted thread.
- If invalid tweets are generated. For example: too much text, too many media items attached to a single tweet.

### How to break it

- Try lots of different types of content.
- Try lots of different types of text, especially in non-English languages. Even better is languages with a different formatting to English For example: RTL languages, CJK (and related) languages.
- URLs have some special treatment, both those typed as text, and those added as links. Try adding URLs to the content in lots of different ways.

## "Fitting the Twitter Format"

Despite tweets not allowing text formatting, there are certain conventions that many folks used. List items usually have a hyphen or asterisk at the start of the line, for example.

### What would a bug look like?

- Blocks having formatting when they shouldn't, or vice-versa.
- Formatting being applied incorrectly.
- Tweets that would otherwise valid, ending up too long due to formatting being applied.

### How to break it

- Particularly for the blocks that apply formatting (List block, Quote block), are you able to make the formatting come out looking wrong?
- For inline links, can add links in a way that cause them to be formatted incorrectly?

## "Useful Feedback... Indicating... Transformations"

While writing a post, there are a few UI elements that provide rapid feedback on how the content will be tweeted. These are:

- The tweet boundary. When content will be split (for example, a long paragraph will be split into two tweets), this indicator shows where the split will happen within the text.
- The tweet divider. This shows between blocks, after the last block within the tweet that the author is currently editing.
- Warning popover. When the author is clearly not focussed elsewhere, a popover map show with the tweet divider, indicating that content in the current tweet may be modified when posting to Twitter.

These elements are deliberately subtle, and should fade into the background when the author is typing.

The Social Previews modal takes the author out of the block editor flow, but shows a more accurate representation of what the thread will look like when posted to Twitter: applying transformations, explicitly splitting content between tweets, and removing content that won't be tweeted.

### What would a bug look like?

- The tweet boundary or divider appearing in the wrong place, or not appearing when they should.
- The popover blocking something you were trying to do.
- The UI not fading out when you're typing.
- The Social Previews UI not showing tweets as they should appear.

### How to break it

- Can you produce content that causes the tweet boundary to appear somewhere other than on a space character? (It may appear as if the letter it appears on is squashed into the next one.)
- The list block has an additional boundary style, which appears at the end of list items. Does it always show on the correct line?
- Can the UI be tricked into not fading when it should, perhaps through alternate input methods?
- The boundaries are calculated through an asynchronous call to the server, which means there's time between when the server call happens, and when it returns, you could've typed more. It tries to not update boundaries on content that has changed since they were calculated, but can you trick into doing that?

## "Tweak... Content"

While this feature aims to avoid locking the author into the rigid structure of a Twitter thread, it's also not unreasonable that they would want to slightly alter content to fit it better into the thread format. For example, given a sentence with 281 characters, it's perfectly reasonable to want to edit it down slightly, rather than having the last word off the sentence split into the next tweet.

### What would a bug look like?

- When tweaking content, finding that the UI hinders progress, rather than helping.
- Finding that the UI gives inaccurate feedback, resulting in unnecessary tweaking.

### How to break it

- Try making different kinds of edits that alter how content is split into tweets.

# Miscellany

- I strongly advise having [twitter.com](https://twitter.com/) open in another tab, it's an easy reference for checking whether suspicious content actually fits into a tweet.
- If you notice something looking odd, it's helpful to take screenshots and copy the broken content first, before trying to reproduce it, or reduce it down to a simple test case. Additionally, taking a moment to note down what steps you performed recently can help make it easier to reproduce multi-step issues.
- For testing the content of threads, remember to check that you're connected to a test Twitter account, rather than your primary account. I fortunately only made that mistake once, with just a short thread. 😉

davemart-in commented:
@pento for folks who don't have a WP.com sandbox, would you be up for recording a quick demo preview of this feature?

pento commented:
@davemart-in:
> for folks who don't have a WP.com sandbox, would you be up for recording a quick demo preview video of this feature?

Yes, that can be done! I'm aiming to get that part deployed early next week, but a little video walkthrough would be helpful as well, I think. I'll get that done on Monday. 🙂

aaronrobertshaw commented on _inc/lib/core-api/wpcom-endpoints/class-wpcom-rest-api-v2-endpoint-tweetstorm-parse.php:
```diff
@@ -0,0 +1,110 @@
+			array(
+				'args'                                  => array(
+					'blocks' => array(
+						'description' => __( 'An array of serialised blocks, and editor-specific block information.', 'jetpack' ),
```
Nitpick: We should tweak this to the American spelling.
```suggestion
						'description' => __( 'An array of serialized blocks, and editor-specific block information.', 'jetpack' ),
```

glendaviesnz commented:
Not sure if it is intentional or not, but if you select to publish a thread and then cancel the publish box, the calls to the thread api continue to run in the background, which makes for a lot of api calls if user continues doing a lot of editing, and a bit hard to tell, but it does seem to impact performance of the editor a fraction while doing this:

![twitter-thread](https://user-images.githubusercontent.com/3629020/93826357-26e75480-fcbb-11ea-92ad-b764cdbf8b93.gif)

Maybe the background parsing of the post should only happen if the right publish panel is open?

glendaviesnz commented:
I wonder if all the api parsing calls should only happen when the `Publish` panel is open ... if a user cancels that and decides to do a lot more editing is it worth running all the parsing still as they type, or just activate it again once the publish panel is open?

pento commented:
> I wonder if all the api parsing calls should only happen when the `Publish` panel is open ... if a user cancels that and decides to do a lot more editing is it worth running all the parsing still as they type, or just activate it again once the publish panel is open?

It does need to do some sort of checking even when the panel is closed, to ensure the tweet boundaries are updated correctly. The current method is fairly brute-force, there are certainly ways it could be optimised:

- Rather than sending the entire post for parsing, just send blocks that will have potentially had their boundaries changed.
- Be a bit smarter about when parsing is needed: if the changes since last parse definitely wouldn't have changed the tweet boundaries, there's no need to re-parse.

I've added these to the [Future Work Overview](https://github.com/Automattic/jetpack/issues/17149) for tracking, but I don't think they're a blocker for launch.

kbrown9 commented:
@pento - PR #17144 has been merged, so the autoloader can load PSR-0 packages now!

aaronrobertshaw commented on extensions/blocks/publicize/store/effects.js:
```diff
@@ -1,32 +1,121 @@
+}
+
+/**
+ * Effect handler which will refreshing the state of the tweets. Tweet refreshes are throttled
```
Nitpick: Just a typo.

```suggestion
 * Effect handler which will refresh the state of the tweets. Tweet refreshes are throttled
```

aaronrobertshaw commented:
I've been able to take a look over the JS side of this PR. I only found one issue related to the Twitter Social Preview which @pento has now [fixed](https://github.com/Automattic/jetpack/pull/16699/commits/d95a91b2067096171c82fbc7e3fb0820b1e2f460). 

* Tried with multiple combinations of different block types, both supported and unsupported
* Tried injecting weird content into supported blocks but couldn't break the publishing
* The editor UI adjustments all seemed to work and were not intrusive
* Works well as far as I was able to test (still feel an extra set of eyes or two would be beneficial)
* Primarily tested using ephemeral site and Jetpack beta plugin

kraftbj reviewed (approved):
YOLO Monday.

Let's do it.

There are a couple of things that I'd like to see this week before release OR, alternatively, marking it as a beta block so we can get a lot more testing while we polish.

0. I like this list https://github.com/Automattic/jetpack/issues/17149 with having the spammy limits added.
1. I think there should be some indicator (sidebar where the preview and setting is?) of what blocks won't convert. It makes a bit of an awkward tweet when chunks are missing because of unsupported blocks and even moreso if the entire content is not supported https://twitter.com/BJGKraft/status/1310681995882778627

I think having some type of "your whole post won't be threaded!" are-you-sure or warning, etc listing what won't work would help prevent a number of support questions on why the thread and the post aren't the same.

pento commented:
> I like this list #17149 with having the spammy limits added.

👍🏻

> I think there should be some indicator (sidebar where the preview and setting is?) of what blocks won't convert. It makes a bit of an awkward tweet when chunks are missing because of unsupported blocks and even moreso if the entire content is not supported [twitter.com/BJGKraft/status/1310681995882778627](https://twitter.com/BJGKraft/status/1310681995882778627)

It does show this, but the UI is pretty cautious about when it shows the popover, since it's just a distraction when you're trying to write.

![](https://user-images.githubusercontent.com/352291/90364190-5b0d9b00-e0a7-11ea-803f-a30a904d765c.png)

> I think having some type of "your whole post won't be threaded!" are-you-sure or warning, etc listing what won't work would help prevent a number of support questions on why the thread and the post aren't the same.

That's an interesting idea, though I wouldn't want it showing all the time. The preview does convey that information, but perhaps it should show in the pre-publish panel, too.

kraftbj commented:
> That's an interesting idea, though I wouldn't want it showing all the time. The preview does convey that information, but perhaps it should show in the pre-publish panel, too.

I didn't see the hover element on the blocks -- which is fine. The pre-publish is where I was expecting there to be the warning. I write dangerously and don't always preview ;-).

jeherve commented:
r214755-wpcom
